[Build] Cannot complete build - Error at MoE Step

### Describe the issue

WSL2 
Cuda 12.4
CuDNN 9.0.0
Latest Cmake 3.29.0-rc3

Last working Commit: 3dfe4a5e6d5075a976516a9c43fe5f92da8614a9

Unable to complete build since above commit. Errors out at MoE step. 





### Urgency

_No response_

### Target platform

WSL2 (Ubuntu 22.04)

### Build script

./build.sh --use_cuda --cudnn_home /usr/include --cuda_home /usr/local/cuda-12.4 --config=Release --build_shared_lib --build_wheel --skip_tests --parallel --compile_no_warning_as_error

### Error / output

Error Messages: 
``` 
In file included from /home/media/onnxruntime/include/onnxruntime/core/common/common.h:37,
                 from /home/media/onnxruntime/include/onnxruntime/core/common/exceptions.h:13,
                 from /home/media/onnxruntime/onnxruntime/core/common/safeint.h:6,
                 from /home/media/onnxruntime/onnxruntime/contrib_ops/cuda/moe/moe.cc:4:
/home/media/onnxruntime/include/onnxruntime/core/common/make_string.h: In instantiation of ‘void onnxruntime::detail::MakeStringImpl(std::ostringstream&, const T&) [with T = gsl::span<const long int>; std::ostringstream = std::__cxx11::basic_ostringstream<char>]’:
/home/media/onnxruntime/include/onnxruntime/core/common/make_string.h:38:17:   required from ‘void onnxruntime::detail::MakeStringImpl(std::ostringstream&, const T&, const Args& ...) [with T = gsl::span<const long int>; Args = {const char*, gsl::span<const long int, 18446744073709551615>}; std::ostringstream = std::__cxx11::basic_ostringstream<char>]’
/home/media/onnxruntime/include/onnxruntime/core/common/make_string.h:39:17:   required from ‘void onnxruntime::detail::MakeStringImpl(std::ostringstream&, const T&, const Args& ...) [with T = const char*; Args = {gsl::span<const long int, 18446744073709551615>, const char*, gsl::span<const long int, 18446744073709551615>}; std::ostringstream = std::__cxx11::basic_ostringstream<char>]’
/home/media/onnxruntime/include/onnxruntime/core/common/make_string.h:46:17:   required from ‘std::string onnxruntime::detail::MakeStringImpl(const Args& ...) [with Args = {const char*, gsl::span<const long int, 18446744073709551615>, const char*, gsl::span<const long int, 18446744073709551615>}; std::string = std::__cxx11::basic_string<char>]’
/home/media/onnxruntime/include/onnxruntime/core/common/make_string.h:93:32:   required from ‘std::string onnxruntime::MakeString(const Args& ...) [with Args = {char [73], gsl::span<const long int, 18446744073709551615>, char [6], gsl::span<const long int, 18446744073709551615>}; std::string = std::__cxx11::basic_string<char>]’
/home/media/onnxruntime/onnxruntime/contrib_ops/cuda/moe/moe_base.h:133:14:   required from here
/home/media/onnxruntime/include/onnxruntime/core/common/make_string.h:33:6: error: no match for ‘operator<<’ (operand types are ‘std::ostringstream’ {aka ‘std::__cxx11::basic_ostringstream<char>’} and ‘const gsl::span<const long int>’)
   33 |   ss << t;
      |   ~~~^~~~
```
```
/home/media/onnxruntime/onnxruntime/contrib_ops/cuda/moe/moe_base.h:133:14:   required from here
/usr/include/c++/12/ostream:754:5: error: no type named ‘type’ in ‘struct std::enable_if<false, void>’
```

### Visual Studio Version

_No response_

### GCC / Compiler Version

12.3.0

--- a/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
+++ b/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
@@ -49,8 +49,6 @@
 
 #include "core/providers/cuda/cu_inc/common.cuh"
 
-using namespace onnxruntime::cuda;
-
 namespace ort_fastertransformer {
 static constexpr int WARP_SIZE = 32;
 

--- a/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
+++ b/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
@@ -49,6 +49,8 @@
 
 #include "core/providers/cuda/cu_inc/common.cuh"
 
+using namespace onnxruntime::cuda;
+
 namespace ort_fastertransformer {
 static constexpr int WARP_SIZE = 32;
 

--- a/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
+++ b/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
@@ -47,8 +47,6 @@
 #include "cub/util_type.cuh"
 #endif
 
-#include "core/providers/cuda/cu_inc/common.cuh"
-
 namespace ort_fastertransformer {
 static constexpr int WARP_SIZE = 32;
 
@@ -658,11 +656,27 @@ inline __device__ float4 operator*(const float4 a, const float4 b) {
   return make_float4(a.x * b.x, a.y * b.y, a.z * b.z, a.w * b.w);
 }
 
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
+    ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
+inline __device__ half operator*(const half a, const half b) {
+  return __float2half(__half2float(a) * __half2float(b));
+}
+
+inline __device__ half2 operator*(const half2 a, const half2 b) {
+  return make_half2(a.x * b.x, a.y * b.y);
+}
+#endif
+
 inline __device__ Half4 operator*(const Half4 a, const Half4 b) {
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
+    ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
   Half4 result;
   result.x = a.x * b.x;
   result.y = a.y * b.y;
   return result;
+#else
+  return Half4{__hmul2(a.x, b.x), __hmul2(a.y, b.y)};
+#endif
 }
 
 }  // anonymous namespace

--- a/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
+++ b/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
@@ -47,6 +47,8 @@
 #include "cub/util_type.cuh"
 #endif
 
+#include "core/providers/cuda/cu_inc/common.cuh"
+
 namespace ort_fastertransformer {
 static constexpr int WARP_SIZE = 32;
 
@@ -656,27 +658,11 @@ inline __device__ float4 operator*(const float4 a, const float4 b) {
   return make_float4(a.x * b.x, a.y * b.y, a.z * b.z, a.w * b.w);
 }
 
-#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
-    ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
-inline __device__ half operator*(const half a, const half b) {
-  return __float2half(__half2float(a) * __half2float(b));
-}
-
-inline __device__ half2 operator*(const half2 a, const half2 b) {
-  return make_half2(a.x * b.x, a.y * b.y);
-}
-#endif
-
 inline __device__ Half4 operator*(const Half4 a, const Half4 b) {
-#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
-    ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
   Half4 result;
   result.x = a.x * b.x;
   result.y = a.y * b.y;
   return result;
-#else
-  return Half4{__hmul2(a.x, b.x), __hmul2(a.y, b.y)};
-#endif
 }
 
 }  // anonymous namespace

--- a/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
+++ b/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
@@ -656,7 +656,8 @@ inline __device__ float4 operator*(const float4 a, const float4 b) {
   return make_float4(a.x * b.x, a.y * b.y, a.z * b.z, a.w * b.w);
 }
 
-#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
+    ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
 inline __device__ half operator*(const half a, const half b) {
   return __float2half(__half2float(a) * __half2float(b));
 }
@@ -667,7 +668,8 @@ inline __device__ half2 operator*(const half2 a, const half2 b) {
 #endif
 
 inline __device__ Half4 operator*(const Half4 a, const Half4 b) {
-#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
+    ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
   Half4 result;
   result.x = a.x * b.x;
   result.y = a.y * b.y;

--- a/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
+++ b/onnxruntime/contrib_ops/cuda/moe/ft_moe/moe_kernel.cu
@@ -656,6 +656,7 @@ inline __device__ float4 operator*(const float4 a, const float4 b) {
   return make_float4(a.x * b.x, a.y * b.y, a.z * b.z, a.w * b.w);
 }
 
+// TODO(wy): use cuda common header and investigate pipeline build issue.
 #if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
     ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))
 inline __device__ half operator*(const half a, const half b) {
@@ -667,6 +668,7 @@ inline __device__ half2 operator*(const half2 a, const half2 b) {
 }
 #endif
 
+// TODO(wy): use cuda common header and investigate pipeline build issue.
 inline __device__ Half4 operator*(const Half4 a, const Half4 b) {
 #if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 530 && \
     ((__CUDACC_VER_MAJOR__ < 12) || ((__CUDACC_VER_MAJOR__ == 12) && (__CUDACC_VER_MINOR__ < 2)))

